### PR TITLE
Fixed in Create Order the update of cart rules when changing adress 

### LINKED ISSUE #21548: BO - Orders page - Vouchers block - Voucher line is not added when a cart rule "Discount for a country selection" is applied
#### Describe the bug

In the BO => Orders => New Order page, the "Discount" line is not added to the Vouchers block when a cart rule "Discount for a country selection" is applied

#### Expected behavior

Explain what you expected to happen instead.

#### Steps to Reproduce

You have at least France and United States countries enabled

1. Go to BO => Catalog => Discounts page
2. Create a cart rule Discount amount with a country selection "The United States"
 image
4. Go to BO => Orders => Add new order page for "John Doe"
5. add some product to the order
6. Select the address "My address (United states country)" 
7. See an error, the Discount line is not added to Voucher Block
8. In the Summary Block => ok

image

**Other steps to reproduce the issue:**

1. Go to BO => Catalog => Discounts page
2. Create a cart rule Discount amount with a country selection "The United States"
image
4. Go to BO => Orders => Add new order page for "John Doe"
5. Select a "My address (United states country)" 
6. add some product to the order
7. The Voucher well added 
8. Try to change the address selected to France
9. See the error "The voucher line is not removed
10. In the Summary block, Total voucher = 0 (well updated)

image

PS1.7.6.8 => ok

#### Additional information

* PrestaShop version: 1770rc1build2
* PHP version: 7.2

### LINKED ISSUE #21549: BO - Order page - The Voucher "Free shipping " value is not updated when changing an address
#### Describe the bug

The Voucher "Free shipping " value is not updated when changing an address
image

#### Steps to Reproduce

1. Create a Free shipping cart rule with a code
2. Enable the Carrier "My light carrier"
3. Go to BO > Add new Order page (For John Doe)
4. Add "demo_11" with quantity = 8 to the cart
5. Apply the discount
4. Change Address
5. see the error the Voucher value is incorrect
image

#### Additional information

* PrestaShop version: 1.7.7.0rc1build2
* PHP version: 7.2

### LINKED ISSUE #22124: BO - Order page - The shipping block is not instantly updated when updating the Address
#### Describe the bug

In the BO > Orders > Add new Order page, The shipping block is not instantly updated when updating the Address.

#### Steps to Reproduce

0. Install PS with English language & France country (The United States country is not displayed by default)
1. Go to BO => Orders => Add new order page for "John Doe", choose him.
2. Select the address "My address (United states country)"
3. add some product to the order
image
4. Try to Edit the Address with France country
5. Save
6. See error
7. The Shipping block is not updated
image
PS: if you change the selected address => it will be ok
image

**Screenshots**

PS1769 => ok https://drive.google.com/file/d/1S4zkzTfZGj456oINqQWjvDu6KjCpoIv1/view?usp=sharing
PS1770stablebuild1 => NOK: https://drive.google.com/file/d/1W4wk-gIL-BOirm3YrrvbWAqLlxIMRIG5/view?usp=sharing

#### Additional information

* PrestaShop version: 1770stablebuild1
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Fixed in Create Order the update of cart rules when changing adress 
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #21548 & Fixes #21549
| How to test?  | Cf https://github.com/PrestaShop/PrestaShop/issues/21548#issue-726345492

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21568)

### CODE REVIEW

**Review by atomiix — DISMISSED** [review]

**Review by sowbiba — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,


I have an error following these steps:
1. Create a cart rule to send a free gift ( no code / no conditions)
image
2. Go to BO > Order page > Add a new order for "John Doe"
3. Add a product to the cart (the cart rule well applied)
4. Try to remove the cart rule
5. see the error
image

---

I found this issue:
1. Go to BO => Catalog => Discounts page
2. Create a cart rule Discount amount with a country selection "The United States"
3. Go to BO => Orders => Add new order page for "John Doe"
4. add some product to the order
5. Select the address "My address (United states country)"
6. The voucher well added => ok
7. try to edit the address selected to Country France
8. see error => the discount Is still displayed

https://drive.google.com/file/d/16aty_EEwaFPO1sOaVBDKLJoGuoEum2Np/view?usp=sharing

Similar issue with other steps:

1. Go to BO => Catalog => Discounts page
2. Create a cart rule Discount amount with a country selection "The United States"
3. Go to BO => Orders => Add new order page for "John Doe"
4. Select  "Mon address (France country)"
5. add some product to the order
6. try to edit the address selected to Country United States
7. see error => the discount is not applied


About this issue: https://github.com/PrestaShop/PrestaShop/issues/21549
If you change the address selected => OK, But if you edit the address => NOK
I attached a screen record
https://drive.google.com/file/d/1D3FdnsPRo_Hbd_1I9lRHuSR6dyeRk9BS/view?usp=sharing

Thanks!

**Comment by Progi1984** [comment]
@khouloudbelguith For the first bug, it's not related to this PR.

But for the second, I'm not sure. I need to dig.

**Comment by Progi1984** [comment]
@khouloudbelguith The first problem is coming from `modules/ps_checkout/ps_checkout.php` in the method `hookActionCartUpdateQuantityBefore` where the variable `$this->context->cart->id` is not checked.

But the second bug is fixed.

**Review by matthieu-rolland — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Scenario/Cart/Calculation/Country/cart_rule_country_specific.feature [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s cart --tags calculation-country-cart-rule-specific
+
+@reset-database-before-feature
+@calculation-country-cart-rule-specific
+Feature: Cart calculation with country specific cart rules
+  As a customer
+  I must be able to have correct cart total when selecting country
+
+  Scenario: Cart with a Product And adress restricted by cart rule
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Cart/Calculation/Country/cart_rule_country_specific.feature]:
```suggestion
  Scenario: Cart with a Product And address restricted by cart rule
```

**Review by jolelievre — DISMISSED** [review]
Small typo but not blocking

File: tests/Integration/Behaviour/Features/Context/CartRuleFeatureContext.php [hunk 2]
```diff
@@ -221,6 +227,25 @@ public function cartRuleNamedIsRestrictedToCarrierNamed($cartRuleName, $carrierN
...
+        $this->countryFeatureContext->checkCountryWithIsoCodeExists($country);
+        $this->cartRules[$cartRuleName]->country_restriction = 1;
+        $this->cartRules[$cartRuleName]->save();
+
+        $idCartRule = (int) $this->cartRules[$cartRuleName]->id;
+        $idCountry = (int) $this->countryFeatureContext->getCountryWithIsoCode($country);
+        Db::getInstance()->execute('
+          INSERT INTO ' . _DB_PREFIX_ . "cart_rule_country(`id_cart_rule`, `id_country`)
+          VALUES('" . $idCartRule . "', '" . $idCountry . "')
+        ");
```
Comment (PierreRambaud) [inline tests/Integration/Behaviour/Features/Context/CartRuleFeatureContext.php]:
```suggestion
        Db::getInstance()->execute(
            'INSERT INTO ' . _DB_PREFIX_ . 'cart_rule_country(`id_cart_rule`, `id_country`) ' .
            'VALUES(' . $idCartRule . ', ' . $idCountry . ')'
        );
```

**Review by jolelievre — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

I found an issue following these steps:

1. Go to BO => Catalog => Discounts page
2. Create a cart rule with a country selection "The United States" with an action "Free gift"
3. Go to BO => Orders => Add new order page for "John Doe"
4. Select "Mon address (France country)"
5. add some product to the order
6. try to edit the address selected to Country United States / or select another address "My address (United states country)"
7. see error => the discount is applied => but the gift is not added to the cart.
image

Similar steps to reproduce the issue

1.1. Go to BO => Catalog => Discounts page
2. Create a cart rule with a country selection "The United States" with an action "Free gift"
3. Go to BO => Orders => Add new order page for "John Doe"
4. Select the address "My address (United states country)"
5. add some product to the order (the discount well updated)
6. try to edit the address selected to Country United States / or select another address "(France country)"
7.see error => the discount is removed => but the gift still displayed
image


Thanks!

**Comment by Progi1984** [comment]
@khouloudbelguith I really think the bug is not linked (and should be managed in another PR). And i recently added the #22124 to this PR. So I think a new check must be done.

**Review by PierreRambaud — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

It is ok :heavy_check_mark: 
https://drive.google.com/file/d/10i2PaeDC84aFAiIncCQEHoZQPW4pbP5T/view?usp=sharing

If this PR is merged, I will create a new issue (regression) about the gift as described here: https://github.com/PrestaShop/PrestaShop/pull/21568#issuecomment-734300305

Thanks!

**Comment by Progi1984** [comment]
@khouloudbelguith Thanks. It's merged

**Comment by khouloudbelguith** [comment]
Hi @Progi1984,

Done: https://github.com/PrestaShop/PrestaShop/issues/22243

Thank you!

**Pull request merged**